pkg/virtual/apiexport/authorizer/content.go

```diff
@@ -0,0 +1,102 @@
+
+	kaudit "k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/client-go/kubernetes"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpauth "github.com/kcp-dev/kcp/pkg/authorization"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+const (
+	VirtualAPIExportNotPermittedReason = "virtual apiexport access not permitted"
+
+	VirtualAPIExportContentAuditPrefix   = "virtual.apiexport.content.authorization.kcp.dev/"
+	VirtualAPIExportContentAuditDecision = VirtualAPIExportContentAuditPrefix + "decision"
+	VirtualAPIExportContentAuditReason   = VirtualAPIExportContentAuditPrefix + "reason"
+
+	VirtualAPIExportMaximumPermissionPolicyAuditPrefix   = "virtual.apiexport.maxpermissionpolicy.authorization.kcp.dev/"
+	VirtualAPIExportMaximumPermissionPolicyAuditDecision = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "decision"
+	VirtualAPIExportMaximumPermissionPolicyAuditReason   = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "reason"
+)
+
+type apiExportsContentAuthorizer struct {
+	newDelegatedAuthorizer func(clusterName string) (authorizer.Authorizer, error)
+}
+
+func NewAPIExportsContentAuthorizer(kubeClusterClient kubernetes.ClusterInterface) authorizer.Authorizer {
+	return &apiExportsContentAuthorizer{
+		newDelegatedAuthorizer: func(clusterName string) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(logicalcluster.New(clusterName), kubeClusterClient)
+		},
+	}
+}
+
+func (a *apiExportsContentAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (dec authorizer.Decision, reason string, err error) {
+	defer func() {
+		auditReason := reason
+		if err != nil {
+			auditReason = fmt.Sprintf("%v, error: %v", reason, err)
+		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportContentAuditDecision, kcpauth.DecisionString(dec),
+			VirtualAPIExportContentAuditReason, auditReason,
+		)
+	}()
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+	}
+
+	apiExportCluster, apiExportName := parts[0], parts[1]
+	authz, err := a.newDelegatedAuthorizer(apiExportCluster)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("error creating delegated authorizer for API export %q: %w", apiExportCluster, err)
+	}
+
+	SARAttributes := authorizer.AttributesRecord{
+		APIGroup:        apisv1alpha1.SchemeGroupVersion.Group,
+		APIVersion:      apisv1alpha1.SchemeGroupVersion.Version,
+		User:            attr.GetUser(),
+		Verb:            attr.GetVerb(),
+		Name:            apiExportName,
+		Resource:        "apiexports",
+		ResourceRequest: true,
+		Subresource:     "content",
+	}
```

Comment:
wondering why this authorizer is needed in the first place. It checks that the APIExport user, i.e. an owner of a permission claim, has content access to the workspace at hand?

Reply:
oh, nvmd. It checks that the "red APIExport B owner" actually has the required access. Makes sense.